docs/proposals/multi-platform.md

```diff
@@ -0,0 +1,477 @@
+
+When cross-compiling on the other hand, `CGO_ENABLED` is set to `0` by default.
+
+To always be safe, run this when compiling statically:
+
+```console
+$ CGO_ENABLED=0 go build -a -installsuffix cgo main-with-net.go
+$ file main-with-net
+main-with-net: ELF 64-bit LSB executable, x86-64, version 1 (SYSV), statically linked, not stripped
+```
+See [golang/go#9344](https://github.com/golang/go/issues/9344) for more details.
+
+### Dynamic cross-compilation
+
+In order to dynamically compile a go binary with `cgo`, we need `gcc` installed at build time.
+
+The only Kubernetes binary that is using C code is the `kubelet`, or in fact `cAdvisor` on which `kubelet` depends. `hyperkube` is also dynamically linked as long as `kubelet` is. We should aim to make `kubelet` statically linked.
+
+The normal `x86_64-linux-gnu` can't cross-compile binaries, so we have to install gcc cross-compilers for every platform. We do this in the [`kube-cross`](https://github.com/kubernetes/kubernetes/blob/master/build/build-image/cross/Dockerfile) image,
+and depend on the [`emdebian.org` repository](https://wiki.debian.org/CrossToolchains). Depending on `emdebian` isn't ideal, so we should consider using the latest `gcc` cross-compiler packages from the `ubuntu` main repositories in the future.
+
+Here's an example when cross-compiling plain C code:
+```c
+// main.c
+#include <stdio.h>
+main()
+{
+  printf("Hello Kubernetes!\n");
+}
+```
+```console
+$ arm-linux-gnueabi-gcc -o main-c main.c
+$ file main-c
+main-c: ELF 32-bit LSB executable, ARM, EABI5 version 1 (SYSV), dynamically linked, 
+    interpreter /lib/ld-linux.so.3, for GNU/Linux 2.6.32, not stripped
+```
+
+And here's an example when cross-compiling `go` and `c`:
+```go
+// main-cgo.go
+package main
+/*
+char* sayhello(void) { return "Hello Kubernetes!"; }
+*/
+import "C"
+import "fmt"
+func main() {
+	fmt.Println(C.GoString(C.sayhello()))
+}
+```
+```console
+$ CGO_ENABLED=1 CC=arm-linux-gnueabi-gcc GOOS=linux GOARCH=arm go build main-cgo.go
+$ file main-cgo
+./main-cgo: ELF 32-bit LSB executable, ARM, EABI5 version 1 (SYSV), dynamically linked, 
+    interpreter /lib/ld-linux.so.3, for GNU/Linux 2.6.32, not stripped
+```
+
+The bad thing with dynamic compilation is that it adds an unnecessary dependency on `glibc` _at runtime_.
+
+### Static compilation with CGO code
+
+Lastly, it's even possible to cross-compile `cgo` code _statically_:
+
+```console
+$ CGO_ENABLED=1 CC=arm-linux-gnueabi-gcc GOARCH=arm go build -ldflags '-extldflags "-static"' main-cgo.go
+$ file main-cgo
+./main-cgo: ELF 32-bit LSB executable, ARM, EABI5 version 1 (SYSV), statically linked,
+    for GNU/Linux 2.6.32, not stripped
+```
+
+This is especially useful if we want to include the binary in a container. 
+If the binary is statically compiled, we may use `busybox` or even `scratch` as the base image.
+This should be the preferred way of compiling binaries that strictly require C code to be a part of it.
```

Comment:
fwiw you can statically cross compile cgo from linux to all other arches, it's not something I am proud of knowing how to do though